### PR TITLE
#6993: GeoCarousel section for geostory

### PR DESCRIPTION
**Pull request by dsuren1**
## Description
This PR adds commits to implement new feature section in Geostory called 'GeoCarousel'

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**
#6993 

**What is the new behavior?**
Refer the issue for detailed behavior (#6993 & https://github.com/geosolutions-it/MapStore2/issues/6993#issuecomment-874031648)

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Review by allyoucanmap — CHANGES_REQUESTED** [review]
Some layout issues:

- [ ] in edit mode the scrollbar in the text box does not work

https://user-images.githubusercontent.com/19175505/125422122-38f759ea-04c0-412d-8362-c320ac8a1b97.mp4

- [ ] Somehow the title is not centered in the cover section
- PR branch:

image

  - dev
  
image

File: package.json [hunk 1]
```diff
@@ -237,6 +237,7 @@
     "react-input-autosize": "1.1.4",
     "react-intersection-observer": "8.24.1",
     "react-intl": "2.3.0",
+    "react-items-carousel": "2.8.0",
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply package.json:240]: react-swipeable doesn't seems to support multiple items per view ([Ref](https://github.com/oliviertassinari/react-swipeable-views/issues/185)). That's why I went with a simple multiple item view carousel lib with swipeable feature in it.

File: web/client/components/geostory/builder/SectionsPreview.jsx [hunk 2]
```diff
@@ -257,7 +305,6 @@ const sectionToItem = ({
                 {
                     onClick: scrollToHandler(id, scrollTo),
                     glyph: 'zoom-to',
-                    // visible: type !== 'immersive',
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/components/geostory/builder/SectionsPreview.jsx:260]: It is a removed line

File: web/client/components/geostory/contents/Column.jsx [hunk 3]
```diff
@@ -47,8 +50,19 @@ export default ({
...
-}) => (
-    <Contents
+}) => {
+    let tools = {
+        [ContentTypes.TEXT]: ['remove'],
+        [MediaTypes.IMAGE]: ['editMedia', size(), 'showCaption', 'remove'],
+        [MediaTypes.MAP]: ['editMedia', 'editMap', size(true), 'showCaption', 'remove'],
+        [ContentTypes.WEBPAGE]: ['editURL', size(true), 'remove'],
+        [MediaTypes.VIDEO]: ['editMedia', 'muted', 'autoplay', 'loop', 'showCaption', 'remove'],
+        ...overrideTools
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/components/geostory/contents/Column.jsx:60]: I agree. It was lame of me to not use overridetools props instead :)  I will modify the overridetools props to a function and modify the paragraph section accordingly

File: web/client/components/geostory/contents/carousel/Carousel.jsx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    const disableAddItem = () => {
+        const {background: {map: { layers = []} = {}} = {}} = find(items, {id: currentContentId}) || {};
+        const {features = []} = find(layers, {id: GEOSTORY}) || {};
+        return !find(features, {contentRefId: currentContentId});
+    };
+
+    return (
+        <>
+            <div className={'ms-section-carousel'}>
+                {editable && <ContentToolbar bsStyle={'primary'} addDisabled={!isMapBackground || disableAddItem()}
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/components/geostory/contents/carousel/Carousel.jsx:105]: Agreed. I just referred the screenshot in the issue and went with primary, sure, I will use default instead.

File: web/client/components/geostory/contents/carousel/Carousel.jsx [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    add={add} editMap={isEditMap} remove={onRemoveAll} tools={['add', 'remove']}/>}
+                <Carousel
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/components/geostory/contents/carousel/Carousel.jsx:107]: I will make the no of items responsive according to container width

File: web/client/components/geostory/contents/ContentToolbar.jsx [hunk 6]
```diff
@@ -113,6 +113,43 @@ const toolButtons = {
         onClick: () => {
             update('showCaption', !showCaption);
         }
+    }),
+    add: ({ editMap = false, addDisabled = false, add = () => {}, bsStyle = 'default' }) => ({
+        glyph: 'plus',
+        visible: true,
+        bsStyle,
+        tooltipId: "geostory.contentToolbar.add",
+        disabled: editMap || addDisabled,
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/components/geostory/contents/ContentToolbar.jsx:122]: This requirement was a feedback requested later during the development. Kindly refer to this comment https://github.com/geosolutions-it/MapStore2/issues/6993#issuecomment-874031648

**Comment by dsuren1** [comment]
> * [ ]  in edit mode the scrollbar in the text box does not work

The scrollbar appears only in view mode for the whole section content, as I thought during edit, it will make the modification space limited when showing a scrollbar in each inner content of section and making the editing process constraints. Kindly let me your thoughts?

File: web/client/components/geostory/builder/SectionsPreview.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (allyoucanmap) [reply web/client/components/geostory/builder/SectionsPreview.jsx:260]: you are right, sorry

**Review by allyoucanmap — APPROVED** [review]
> The scrollbar appears only in view mode for the whole section content, as I thought during edit, it will make the modification space limited when showing a scrollbar in each inner content of section and making the editing process constraints. Kindly let me your thoughts?

In my opinion the layout of sections should be consistent between edit and view mode as much as possible. If we allow this section to extend we get this behaviour that should be avoided for this kind of section:


https://user-images.githubusercontent.com/19175505/125750055-4826b377-74ff-4be4-a804-59bf1d835f26.mp4

We will fix this in a separated issue with the following improvements:

- [ ] scrollbar also in edit mode (maybe we need to introduce a different class because now is using the immersive one)

image


- [ ] if the title is missing the popup should not be visible

image

- [ ] improve the draw marker UX with some message to explain that a user need to click on the map

image

**Comment by allyoucanmap** [comment]
please @dsuren1 backport this improvement on the branch https://github.com/geosolutions-it/MapStore2/tree/geonode-mapstore-client-2.1.x

**Pull request merged**